Dune 3.4+ broken in esy on Windows

<!-- Thank you for filing an issue to help us improve Dune! -->
## Expected Behavior
Dune 3.4+ should build inside esy on Windows.

## Actual Behavior
Dune fails to build with this log:
```
# esy-build-package: building: @opam/dune@opam:3.4.1
# esy-build-package: pwd: C:\Users\grain\.esy\3\b\opam__s__dune-opam__c__3.4.1-4a143223
# esy-build-package: running: "ocaml" "bootstrap.ml" "-j" "4"
ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot unix.cma boot/libs.ml boot/duneboot.ml
.\.duneboot.exe -j 4
cd _boot && C:\Users\grain\.esy\3_\i\ocaml-4.14.0-cd77a8d9\bin\ocamlopt.opt.exe -o ..\dune.exe -g -I +threads unix.cmxa threads.cmxa -no-alias-deps -w -49-6 readdir.c wait3_stubs.c xdg_stubs.c spawn_stubs.c dune_stats_stubs.c fsevents_stubs.c inotify_stubs.c pthread_chdir_stubs.c -args mods_list -cclib -lshell32 -cclib -lole32 -cclib -luuid -alert -unstable
otherlibs/xdg\xdg_stubs.c: In function 'dune_xdg__get_known_folder_path':
otherlibs/xdg\xdg_stubs.c:35:9: warning: implicit declaration of function 'SHGetKnownFolderPath'; did you mean 'SHGetFolderPath'? [-Wimplicit-function-declaration]
otherlibs/xdg\xdg_stubs.c:41:38: error: 'WC_ERR_INVALID_CHARS' undeclared (first use in this function); did you mean 'MB_ERR_INVALID_CHARS'?
otherlibs/xdg\xdg_stubs.c:41:38: note: each undeclared identifier is reported only once for each function it appears in
error: command failed: "ocaml" "bootstrap.ml" "-j" "4"
```

The issues seem to be `SHGetKnownFolderPath` and `WC_ERR_INVALID_CHARS` in the xdg stubs.

## Reproduction

<!-- Otherwise, please detail how the behavior can be reproduced -->

1. On Windows, create an esy.json with the following:
```json
{
    "dependencies": {
        "ocaml": "4.14.0",
        "@opam/dune": "3.4.1"
    },
    "esy": {
        "build": [
            "dune build -p #{self.name}"
        ]
    }
}
```
1. Try to build the project by running `esy`
1. You'll see the same error as posted above

## Specifications

- Version of `dune` (output of `dune --version`): 3.4.1
- Version of `ocaml` (output of `ocamlc --version`) 4.14.0
- Operating system (distribution and version): Windows (Server 2019)

<!-- remove the following if not used -->
## Additional information 

- Link to gist with verbose output (run `dune` with the `--verbose` flag):

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -36,6 +36,8 @@
 - Support `(foreign_objects ...)` field in `(executable ...)` and `(library
   ...)` stanzas (#6084, fixes #4129, @gridbugs)
 
+- Fix compilation of Dune under esy on Windows (#6109, fixes #6098, @nojb)
+
 3.4.1 (26-07-2022)
 ------------------
 

--- a/otherlibs/xdg/xdg_stubs.c
+++ b/otherlibs/xdg/xdg_stubs.c
@@ -5,6 +5,11 @@
 
 #ifdef _WIN32
 
+/*  Windows Vista functions enabled */
+
+#undef _WIN32_WINNT
+#define _WIN32_WINNT 0x0600
+
 #include <Windows.h>
 #include <Knownfolders.h>
 #include <Shlobj.h>

